src/Eccube/Controller/Admin/Shipping/EditController.php

```diff
@@ -124,47 +150,65 @@ public function edit(Request $request, $id = null)
-                    }
-                // no break
-                case 'register':
-
-                    log_info('出荷登録開始', array($TargetShipping->getId()));
-                    // TODO 在庫の有無や販売制限数のチェックなども行う必要があるため、完了処理もcaluclatorのように抽象化できないか検討する.
-                    if ($form->isValid()) {
-                        // TODO 後続にある会員情報の更新のように、完了処理もcaluclatorのように抽象化できないか検討する.
-
-                        // 画面上で削除された明細をremove
-                        foreach ($OriginalOrderItems as $OrderItem) {
-                            if (false === $TargetShipping->getOrderItems()->contains($OrderItem)) {
-                                $OrderItem->setShipping(null);
-                            }
-                        }
-
-                        foreach ($TargetShipping->getOrderItems() as $OrderItem) {
-                            $OrderItem->setShipping($TargetShipping);
-                        }
-                        $this->entityManager->persist($TargetShipping);
-                        $this->entityManager->flush();
+            log_info('出荷登録開始', array($TargetShipping->getId()));
+            // TODO 在庫の有無や販売制限数のチェックなども行う必要があるため、完了処理もcaluclatorのように抽象化できないか検討する.
+            // TODO 後続にある会員情報の更新のように、完了処理もcaluclatorのように抽象化できないか検討する.
+            // 画面上で削除された明細をremove
+            foreach ($OriginalOrderItems as $OrderItem) {
+                if (false === $TargetShipping->getOrderItems()->contains($OrderItem)) {
+                    $OrderItem->setShipping(null);
+                }
+            }
 
-                        $this->addSuccess('admin.shipping.edit.save.complete', 'admin');
-                        $this->addInfo('admin.shipping.edit.save.info', 'admin');
-                        log_info('出荷登録完了', array($TargetShipping->getId()));
+            foreach ($TargetShipping->getOrderItems() as $OrderItem) {
+                $OrderItem->setShipping($TargetShipping);
+            }
 
-                        return $this->redirectToRoute('admin_shipping_edit', array('id' => $TargetShipping->getId()));
+            $OriginShippingStatus = $OriginShipping->getShippingStatus();
+            $TargetShippingStatus = $TargetShipping->getShippingStatus();
+
+            // 出荷ステータス変更時の処理
+            if ($TargetShippingStatus !== null
+             && $TargetShippingStatus->getId() == ShippingStatus::SHIPPED) {
+                // 「出荷済み」にステータスが変更された場合
+                if ($OriginShippingStatus === null
+                 || $OriginShippingStatus->getId() != $TargetShippingStatus->getId()) {
+                    // 出荷日時を更新
+                    $TargetShipping->setShippingDate(new \DateTime());
+                    // 出荷メールを送信
+                    if ($form->get('notify_email')->getData()) {
+                        try {
+                            $this->mailService->sendShippingNotifyMail(
+                              $TargetShipping
+                            );
+                        } catch (\Twig_Error $e) {
```

Comment:
`\Twig_Error` のみでは不十分では？？
少なくとも、 Swift_Mailer の Exception が throw される可能性があります。

Reply:
Hi @nanasess , should I change `\Twig_Error` to `\Exception` to cover all? Because I can not determine what Swift_Mailer Exception will be thrown on `$this->mailService->sendShippingNotifyMail`

Ref: \Swift_Mailer::send
```
public function send(Swift_Mime_SimpleMessage $message, &$failedRecipients = null)
    {
        $failedRecipients = (array) $failedRecipients;

        if (!$this->transport->isStarted()) {
            $this->transport->start();
        }

        $sent = 0;

        try {
            $sent = $this->transport->send($message, $failedRecipients);
        } catch (Swift_RfcComplianceException $e) {
            foreach ($message->getTo() as $address => $name) {
                $failedRecipients[] = $address;
            }
        }

        return $sent;
    }
```

Reply:
Got it, thank you